<?xml version="1.0" encoding="UTF-8"?>
<launch>

    <arg name="plot_cost" default="true"/>

    <arg name="description_file" default="$(find mppi_pole_cart)/resources/pendulum_cart.urdf"/>
    <arg name="ghost_description_file" default="$(find mppi_pole_cart)/resources/pendulum_cart_ghost.urdf"/>

    <!-- TF prefix -->
    <arg name="tf_prefix" default=""/>
    <!-- Load robot description -->
    <param name="robot_description" command="$(find xacro)/xacro $(arg description_file)"/>
    <param name="ghost_description" command="$(find xacro)/xacro $(arg ghost_description_file)"/>

    <!--  Run the test node-->
    <node name="pole_cart_visualizer" type="pole_cart_visualizer" pkg="mppi_pole_cart" output="screen"
          launch-prefix="gdb -ex run --args">
        <rosparam command="load" file="$(find mppi_pole_cart)/config/params.yaml"/>
        <param name="config_file" value="$(find mppi_pole_cart)/config/params.yaml"/>
    </node>

    <!--  Run the plotter node-->
    <node name="plotter" type="plot_info.py" pkg="mppi_ros" output="screen"/>

    <!-- Publish robot state (joints) as TF -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen">
        <param name="publish_frequency" value="100"/>
        <param name="use_tf_static"     value="true"/>
        <param name="tf_prefix"         value="$(arg tf_prefix)"/>
        <remap from="robot_description" to="robot_description"/>
        <remap from="joint_states"      to="/joint_states"/>
    </node>

<!--    Ghost state for rollouts visualization-->
    <node name="robot_ghost_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen">
        <param name="publish_frequency" value="100"/>
        <param name="use_tf_static"     value="true"/>
        <param name="tf_prefix"         value="$(arg tf_prefix)"/>
        <remap from="robot_description" to="ghost_description"/>
        <remap from="joint_states"      to="/joint_ghost_states"/>
    </node>

    <!-- Open Rviz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find mppi_pole_cart)/config/config.rviz" output="screen"/>

    <group if="$(arg plot_cost)">
        <arg name="mp_config" default="$(find mppi_pole_cart)/resources/rqt_multiplot.xml"/>
        <node name="rqt_multiplot" pkg="rqt_multiplot" type="rqt_multiplot" args="--multiplot-config $(arg mp_config)"
              output="screen"/>
    </group>
</launch>
